Add flake attempts to the flaky tests and include known flakes pattern detection

mpryc opened the pull request:
Add flake attempts to the flaky tests and include known flakes pattern detection
    
This adds attempts to the flaky tests, which are caused by known issues.

The known flake pattern detection allows us to specify which flakes are the ones on which we will retry.

Also adds ivenessProbe and startupProbe to the sample-applications. [More info](https://github.com/openshift/oadp-operator/pull/1270#issue-2084844294)

mpryc commented:
Note this is based on another PR which needs to be merged first #1277

mpryc commented:
/retest

mateusoliveira43 commented on tests/e2e/backup_restore_suite_test.go:
```diff
@@ -268,80 +264,79 @@ var _ = Describe("Backup and restore tests", func() {
 			runBackupAndRestore(brCase, expectedErr, updateLastBRcase, updateLastInstallTime)
 		},
-		Entry("MySQL application CSI", BackupRestoreCase{
+		Entry("MySQL application CSI", FlakeAttempts(3), BackupRestoreCase{
```
nice, ginkgo already has this implemented

I would put number of trials as a variable, so when running locally, we can set to zero (and add docs about it)

mateusoliveira43 replied:
ah, and also add a comment about why we are retrying (link to issues for example)

mpryc replied:
Yes. Found out that Gingko has such decorator :)

Ideally if we have flake attempts on not entire test runs but for example for backup/restore/compare separated, but don't want to make such radical change.

The comment is in the git commit log. I don't think comment here is right as we don't know if those are the only flakes here. We know 2 for sure that are already identified.

mpryc replied:
@mateusoliveira43 added variable `flakeAttempts`

mateusoliveira43 reviewed (changes requested):
Also add `FlakeAttempts` decorator to must-gather suite

mpryc commented:
> Also add `FlakeAttempts` decorator to must-gather suite

Done.

mpryc commented:
/retest-required

mpryc commented:
/retest

mateusoliveira43 reviewed (changes requested):
run `make fmt`

mateusoliveira43 reviewed (changes requested):
tested locally and "MySQL application two Vol CSI" failed (3 times in a row)

looking a periodic run https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-oadp-operator-master-4.14-e2e-test-aws-periodic/1749643186097098752/artifacts/e2e-test-aws-periodic/e2e/build-log.txt I assume we need to make sure volume data is in sync with `/tmp/log/`

Since this PR seems to make CI a lot more stable, would add back the `AppReadyDelay` for "MySQL application two Vol CSI", and try later to remove this hardcoded wait

mpryc commented:
> tested locally and "MySQL application two Vol CSI" failed (3 times in a row)
> 
> looking a periodic run https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-oadp-operator-master-4.14-e2e-test-aws-periodic/1749643186097098752/artifacts/e2e-test-aws-periodic/e2e/build-log.txt I assume we need to make sure volume data is in sync with `/tmp/log/`
> 
> Since this PR seems to make CI a lot more stable, would add back the `AppReadyDelay` for "MySQL application two Vol CSI", and try later to remove this hardcoded wait

I also observed the AppReadyDelay is kind of needed at this time, even more I think we should add this to all cases and remove in the future with correct fix

mateusoliveira43 reviewed (approved):
If CI passes without us telling it to re run, I am happy :smile:

weshayutin commented:
/hold 
for ci discussion

mateusoliveira43 commented:
please do not tell CI to re run 

notes for discussion:
- one of the tests failures was infrastructure, the other one was a route problem, that failed the 3 attempts

- the others 4 successful tests, 2 of them did not had retries, 1 of them had 1 retry and another had 2 retries (but in 2 different tests)

kaovilai commented:
<img width="975" alt="image" src="https://github.com/openshift/oadp-operator/assets/11228024/3b8ea5ea-72e6-44d1-952c-8dd11c49bceb">
lol

mpryc commented:
Ah so whenever I add the log msg to the git and it references other issue in there it's visible on other PR. That's how you got famous ! :)

mateusoliveira43 commented on tests/e2e/backup_restore_suite_test.go:
```diff
@@ -261,30 +271,39 @@ var _ = Describe("Backup and restore tests", func() {
 	DescribeTable("Backup and restore applications",
 		func(brCase BackupRestoreCase, expectedErr error) {
+			if CurrentSpecReport().NumAttempts > 1 && !knownFlake {
+				Fail("No known FLAKE found in a subsequent run, marking test as failed.")
```
shouldn't be "... in the previous run..."?

mpryc replied:
yes

mpryc replied:
resolved.

mateusoliveira43 commented on tests/e2e/e2e_suite_test.go:
```diff
@@ -93,6 +104,8 @@ var runTimeClientForSuiteRun client.Client
 var veleroClientForSuiteRun veleroClientset.Interface
 var csiClientForSuiteRun *snapshotv1client.Clientset
 var dpaCR *DpaCustomResource
+var knownFlake bool
```
since we are not adding this to must-gather tests, only backup/restore ones, wouldn't be better to define these variables there?

mpryc replied:
Currently they are only for backup restore, but I can't say for sure it's only there. We really want to have generic way of detecting flake and the problem may be in the future somewhere else where we want also to use this technique.

mpryc commented:
/hold I think there is small nit in the logic, will need to fix it.

weshayutin commented:
If someone finds a FLAKE in the logs where the test IS reexecuted, please post.  Looking through I did not see one :)

mpryc commented:
> If someone finds a FLAKE in the logs where the test IS reexecuted, please post. Looking through I did not see one :)

There was, but for some reason it wasn't found in the flake detection, so I need to understand why:
```
Failed to check and update snapshot content: failed to remove VolumeSnapshotBeingCreated annotation on the content snapcontent-

```

mpryc commented:
/unhold

mpryc commented:
1. Added `describeRestore` logs to also be included in the flake detection.
2. Added as well flake detection in the must gather as I observed on my box that it also flaked with the test on the s3 500 error, so we should have it there, which initially I though it's not required.

mateusoliveira43 commented on tests/e2e/backup_restore_suite_test.go:
```diff
@@ -261,30 +275,39 @@ var _ = Describe("Backup and restore tests", func() {
 			ApplicationNamespace: "mysql-persistent",
 			Name:                 "mysql-csi-e2e",
 			BackupRestoreType:    CSI,
+			AppReadyDelay:        30 * time.Second,
```
why add this to mongo and mysql csi tests?

mpryc replied:
Any CSI from what I understand is affected with this concurrency problem, so that's why.

mateusoliveira43 replied:
were you able to see them fail without timeout?

mpryc replied:
Here is an example:

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_oadp-operator/1253/pull-ci-openshift-oadp-operator-master-4.13-e2e-test-aws/1744307381291978752/build-log.txt

mateusoliveira43 replied:
the example https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_oadp-operator/1253/pull-ci-openshift-oadp-operator-master-4.13-e2e-test-aws/1744307381291978752/artifacts/e2e-test-aws/e2e/build-log.txt already has delay, but even so fails
```
2024/01/08 11:18:32 Sleeping for 30s to allow application to be ready for case mysql-csi-e2e
2024/01/08 11:19:02 Creating backup mysql-csi-e2e-8f06ebf7-ae17-11ee-adb0-0a580a80be46 for case mysql-csi-e2e
2024/01/08 11:19:13 backup phase: InProgress
2024/01/08 11:19:23 backup phase: WaitingForPluginOperations
2024/01/08 11:19:33 backup phase: Completed
...
2024/01/08 11:19:34 Backup for case mysql-csi-e2e succeeded
2024/01/08 11:19:34 waiting for volume snapshot contents snapcontent-601d92f1-1f77-418e-a8eb-4997c53138a7 to be ready
2024/01/08 11:19:34 Uninstalling app for case mysql-csi-e2e
2024/01/08 11:20:20 Creating restore mysql-csi-e2e-8f06f43c-ae17-11ee-adb0-0a580a80be46 for case mysql-csi-e2e
2024/01/08 11:20:20 restore phase: InProgress
2024/01/08 11:20:30 restore phase: Completed
...
2024/01/08 11:20:31 Running dc-post-restore.sh script.
2024/01/08 11:20:31 Running post restore script for mysql-csi-e2e-8f06f43c-ae17-11ee-adb0-0a580a80be46
2024/01/08 11:20:32 command: /usr/bin/bash /go/src/github.com/openshift/oadp-operator/docs/scripts/dc-post-restore.sh mysql-csi-e2e-8f06f43c-ae17-11ee-adb0-0a580a80be46
2024/01/08 11:20:32 stdout:
restore: mysql-csi-e2e-8f06f43c-ae17-11ee-adb0-0a580a80be46
label:   mysql-csi-e2e-8f06f43c-ae17-11ee-adb0-0a580a80be46
Deleting disconnected restore pods
No resources found
2024/01/08 11:20:32 stderr:
2024/01/08 11:20:32 err: <nil>
2024/01/08 11:20:32 Pod mysql-56cbbc4c98-8k5dl not yet succeeded: phase is Pending
2024/01/08 11:20:37 Pod mysql-56cbbc4c98-8k5dl not yet succeeded: phase is Pending
2024/01/08 11:20:42 Pod todolist-1-b9lgq not yet succeeded: phase is Pending
...
2024/01/08 11:29:30 Pod todolist-1-b9lgq not yet succeeded: phase is Pending
  [FAILED] Timed out after 540.001s.
  Expected
      <bool>: false
  to be true
```
what problems can this delay fix for non-twovol case?

mpryc replied:
I believe the following one:
```
could not restore, VolumeSnapshotContent "snapcontent-" already exists.
```

mateusoliveira43 commented on tests/e2e/must-gather_suite_test.go:
```diff
@@ -23,11 +23,18 @@ var _ = Describe("Backup and restore tests with must-gather", func() {
 	}
 
 	var _ = AfterEach(func(ctx SpecContext) {
+		knownFlake = false
```
NO we had this issue in the past of duplicating code from backup/restore tests to must-gather tests. Add code to `tearDownBackupAndRestore` and `runBackupAndRestore` functions

mpryc replied:
Fixed.

mateusoliveira43 replied:
still has code outside `runBackupAndRestore` function

mpryc replied:
Which ?

The following one? I don't like the following to be outside of DescribeTable so it's a first thing it's happening

```
			if CurrentSpecReport().NumAttempts > 1 && !knownFlake {
				Fail("No known FLAKE found in a previous run, marking test as failed.")
			}

```

mateusoliveira43 commented on tests/e2e/backup_restore_suite_test.go:
```diff
@@ -154,8 +154,15 @@ func runBackupAndRestore(brCase BackupRestoreCase, expectedErr error, updateLast
+	backupLogs := BackupLogs(kubernetesClientForSuiteRun, dpaCR.Client, backup)
+	backupErrorLogs := BackupErrorLogs(kubernetesClientForSuiteRun, dpaCR.Client, backup)
+	accumulatedTestLogs = append(accumulatedTestLogs, describeBackup)
+	accumulatedTestLogs = append(accumulatedTestLogs, backupLogs)
```
can't we append all items at once? and why not also append ErrorLogs?

mpryc replied:
Yes we can, there is `...` option, however I prefer to be explicit here:

```
accumulatedTestLogs = append(accumulatedTestLogs, describeBackup, backupLogs...)
```

mateusoliveira43 reviewed (approved):
Checked the failed job, I think was the route problem (but only happened in one test, of that suite, among all suites, in other words, 1/66 ~= 1% failure). I would say its good to merge

mpryc commented:
/retest

mpryc commented:
@mateusoliveira43 I have not seen that error before. Looks like block device pod is unable to come up in a 9min time limit, and thats why DB pod can not be reached even from within itself. However I don't see from the logs anything that would be easy to say why it happened.

mateusoliveira43 commented:
should we cherry-pick this one?
